File: project/drivers/driver/swift/Tests/ScratchBirdTests/AuthBootstrapContractTests.swift
// Copyright (c) 2026 ScratchBird Software Inc.
//
// This Source Code Form is subject to the terms of the Mozilla Public
// License, v. 2.0. If a copy of the MPL was not distributed with this
// file, You can obtain one at https://mozilla.org/MPL/2.0/.
//
// SPDX-License-Identifier: MPL-2.0

import Foundation
import XCTest
@testable import ScratchBird

#if canImport(Glibc)
import Glibc
private func testSocket(_ domain: Int32, _ type: Int32, _ proto: Int32) -> Int32 {
    Glibc.socket(domain, type, proto)
}
private func testBind(_ fd: Int32, _ addr: UnsafePointer<sockaddr>, _ len: socklen_t) -> Int32 {
    Glibc.bind(fd, addr, len)
}
private func testListen(_ fd: Int32, _ backlog: Int32) -> Int32 {
    Glibc.listen(fd, backlog)
}
private func testAccept(_ fd: Int32, _ addr: UnsafeMutablePointer<sockaddr>?, _ len: UnsafeMutablePointer<socklen_t>?) -> Int32 {
    Glibc.accept(fd, addr, len)
}
private func testClose(_ fd: Int32) -> Int32 {
    Glibc.close(fd)
}
private func testSend(_ fd: Int32, _ buf: UnsafeRawPointer?, _ len: Int, _ flags: Int32) -> Int {
    Glibc.send(fd, buf, len, flags)
}
private func testRecv(_ fd: Int32, _ buf: UnsafeMutableRawPointer?, _ len: Int, _ flags: Int32) -> Int {
    Glibc.recv(fd, buf, len, flags)
}
private func testGetSockName(_ fd: Int32, _ addr: UnsafeMutablePointer<sockaddr>?, _ len: UnsafeMutablePointer<socklen_t>?) -> Int32 {
    Glibc.getsockname(fd, addr, len)
}
private let socketStreamType = Int32(SOCK_STREAM.rawValue)
#else
import Darwin
private func testSocket(_ domain: Int32, _ type: Int32, _ proto: Int32) -> Int32 {
    Darwin.socket(domain, type, proto)
}
private func testBind(_ fd: Int32, _ addr: UnsafePointer<sockaddr>, _ len: socklen_t) -> Int32 {
    Darwin.bind(fd, addr, len)
}
private func testListen(_ fd: Int32, _ backlog: Int32) -> Int32 {
    Darwin.listen(fd, backlog)
}
private func testAccept(_ fd: Int32, _ addr: UnsafeMutablePointer<sockaddr>?, _ len: UnsafeMutablePointer<socklen_t>?) -> Int32 {
    Darwin.accept(fd, addr, len)
}
private func testClose(_ fd: Int32) -> Int32 {
    Darwin.close(fd)
}
private func testSend(_ fd: Int32, _ buf: UnsafeRawPointer?, _ len: Int, _ flags: Int32) -> Int {
    Darwin.send(fd, buf, len, flags)
}
private func testRecv(_ fd: Int32, _ buf: UnsafeMutableRawPointer?, _ len: Int, _ flags: Int32) -> Int {
    Darwin.recv(fd, buf, len, flags)
}
private func testGetSockName(_ fd: Int32, _ addr: UnsafeMutablePointer<sockaddr>?, _ len: UnsafeMutablePointer<socklen_t>?) -> Int32 {
    Darwin.getsockname(fd, addr, len)
}
private let socketStreamType = Int32(SOCK_STREAM)
#endif

private let testManagerProtocolMagic: UInt32 = 0x42444253
private let testManagerProtocolVersion: UInt16 = 0x0101
private let testManagerHeaderSize = 12
private let testMcpMsgStatusResponse: UInt8 = 0x64
private let testMcpMsgHello: UInt8 = 0x65
private let testMcpMsgAuthStart: UInt8 = 0x66
private let testMcpMsgAuthChallenge: UInt8 = 0x12

final class AuthBootstrapContractTests: XCTestCase {
    func testProbeAuthSurfaceDirectScramSha512() async throws {
        let server = try LoopbackAuthServer { clientFd in
            let startup = try readProtocolMessage(from: clientFd)
            XCTAssertEqual(startup.header.type, .startup)
            try writeProtocolMessage(
                to: clientFd,
                type: .authRequest,
                payload: makeAuthRequestPayload(method: authScramSha512Method)
            )
        }

        let probe = try await ScratchBirdConnection.probeAuthSurface(
            ScratchBirdConfig(
                host: "127.0.0.1",
                port: server.port,
                database: "db",
                user: "user",
                sslmode: "disable"
            )
        )
        try server.wait()

        XCTAssertTrue(probe.reachable)
        XCTAssertEqual(probe.ingressMode, "direct")
        XCTAssertEqual(probe.requiredMethod, "SCRAM_SHA_512")
        XCTAssertEqual(probe.requiredPluginMethodId, "scratchbird.auth.scram_sha_512")
        XCTAssertEqual(probe.admittedMethods.count, 1)
        XCTAssertEqual(probe.admittedMethods.first?.wireMethod, "SCRAM_SHA_512")
        XCTAssertTrue(probe.admittedMethods.first?.executableLocally ?? false)
        XCTAssertTrue(probe.additionalContinuationPossible)
    }

    func testProbeAuthSurfaceManagerToken() async throws {
        let server = try LoopbackAuthServer { clientFd in
            let hello = try readManagerFrame(from: clientFd)
            XCTAssertEqual(hello.type, testMcpMsgHello)
            try writeManagerFrame(to: clientFd, type: testMcpMsgStatusResponse, payload: Data())

            let authStart = try readManagerFrame(from: clientFd)
            XCTAssertEqual(authStart.type, testMcpMsgAuthStart)
            try writeManagerFrame(to: clientFd, type: testMcpMsgAuthChallenge, payload: Data())
        }

        let probe = try await ScratchBirdConnection.probeAuthSurface(
            ScratchBirdConfig(
                host: "127.0.0.1",
                port: server.port,
                frontDoorMode: "manager_proxy",
                database: "db",
                user: "admin",
                sslmode: "disable"
            )
        )
        try server.wait()

        XCTAssertTrue(probe.reachable)
        XCTAssertEqual(probe.ingressMode, "manager_proxy")
        XCTAssertEqual(probe.requiredMethod, "TOKEN")
        XCTAssertEqual(probe.requiredPluginMethodId, "scratchbird.auth.authkey_token")
        XCTAssertEqual(probe.admittedMethods.first?.wireMethod, "TOKEN")
        XCTAssertTrue(probe.additionalContinuationPossible)
    }

    func testConnectTracksResolvedScramSha512AuthContext() async throws {
        let password = "secret"
        let server = try LoopbackAuthServer { clientFd in
            _ = try readProtocolMessage(from: clientFd)
            try writeProtocolMessage(
                to: clientFd,
                type: .authRequest,
                payload: makeAuthRequestPayload(method: authScramSha512Method)
            )

            let firstResponse = try readProtocolMessage(from: clientFd)
            XCTAssertEqual(firstResponse.header.type, .authResponse)
            let clientFirst = String(data: firstResponse.payload, encoding: .utf8) ?? ""
            let nonce = extractScramNonce(fromClientFirst: clientFirst)
            let salt = Data("swift-salt".utf8).base64EncodedString()
            let serverFirst = "r=\(nonce)server,s=\(salt),i=4096"
            try writeProtocolMessage(
                to: clientFd,
                type: .authContinue,
                payload: makeAuthContinuePayload(method: authScramSha512Method, stage: 0, value: serverFirst)
            )

            let finalResponse = try readProtocolMessage(from: clientFd)
            XCTAssertEqual(finalResponse.header.type, .authResponse)
            XCTAssertFalse(finalResponse.payload.isEmpty)

            try writeProtocolMessage(
                to: clientFd,
                type: .authOk,
                payload: makeAuthOkPayload(serverInfo: "")
            )
            try writeProtocolMessage(to: clientFd, type: .ready, payload: Data())
        }

        let connection = try await ScratchBirdConnection.connect(
            ScratchBirdConfig(
                host: "127.0.0.1",
                port: server.port,
                database: "db",
                user: "user",
                password: password,
                sslmode: "disable"
            )
        )
        try server.wait()

        let context = connection.getResolvedAuthContext()
        XCTAssertEqual(context.ingressMode, "direct")
        XCTAssertEqual(context.resolvedAuthMethod, "SCRAM_SHA_512")
        XCTAssertEqual(context.resolvedAuthPluginId, "scratchbird.auth.scram_sha_512")
        XCTAssertFalse(context.managerAuthenticated)
        XCTAssertTrue(context.attached)

        try await connection.close()
        XCTAssertFalse(connection.getResolvedAuthContext().attached)
    }

    func testConnectTracksResolvedTokenAuthContext() async throws {
        let token = "demo-token"
        let server = try LoopbackAuthServer { clientFd in
            _ = try readProtocolMessage(from: clientFd)
            try writeProtocolMessage(
                to: clientFd,
                type: .authRequest,
                payload: makeAuthRequestPayload(method: authTokenMethod)
            )

            let response = try readProtocolMessage(from: clientFd)
            XCTAssertEqual(response.header.type, .authResponse)
            XCTAssertEqual(String(data: response.payload, encoding: .utf8), token)

            try writeProtocolMessage(
                to: clientFd,
                type: .authOk,
                payload: makeAuthOkPayload(serverInfo: "")
            )
            try writeProtocolMessage(to: clientFd, type: .ready, payload: Data())
        }

        let connection = try await ScratchBirdConnection.connect(
            ScratchBirdConfig(
                host: "127.0.0.1",
                port: server.port,
                database: "db",
                user: "user",
                sslmode: "disable",
                authToken: token
            )
        )
        try server.wait()

        let context = connection.getResolvedAuthContext()
        XCTAssertEqual(context.resolvedAuthMethod, "TOKEN")
        XCTAssertEqual(context.resolvedAuthPluginId, "scratchbird.auth.authkey_token")
        XCTAssertTrue(context.attached)

        try await connection.close()
    }

    func testConnectFailsClosedForPeerAuth() async throws {
        let server = try LoopbackAuthServer { clientFd in
            _ = try readProtocolMessage(from: clientFd)
            try writeProtocolMessage(
                to: clientFd,
                type: .authRequest,
                payload: makeAuthRequestPayload(method: authPeerMethod)
            )
        }

        do {
            _ = try await ScratchBirdConnection.connect(
                ScratchBirdConfig(
                    host: "127.0.0.1",
                    port: server.port,
                    database: "db",
                    user: "user",
                    sslmode: "disable"
                )
            )
            XCTFail("Expected PEER auth to fail closed")
        } catch let error as ScratchBirdNotSupportedException {
            XCTAssertEqual(error.sqlState, "0A000")
            XCTAssertTrue(error.message.contains("PEER"))
        }

        try server.wait()
    }
}

private final class LoopbackAuthServer {
    let port: Int
    private let listenFd: Int32
    private let group = DispatchGroup()
    private var scenarioError: Error?

    init(handler: @escaping (Int32) throws -> Void) throws {
        let listenFd = testSocket(AF_INET, socketStreamType, 0)
        if listenFd < 0 {
            throw NSError(domain: "ScratchBirdTests", code: -1, userInfo: [NSLocalizedDescriptionKey: "socket() failed"])
        }
        self.listenFd = listenFd

        var reuse: Int32 = 1
        _ = withUnsafePointer(to: &reuse) {
            $0.withMemoryRebound(to: UInt8.self, capacity: MemoryLayout<Int32>.size) {
                setsockopt(listenFd, SOL_SOCKET, SO_REUSEADDR, $0, socklen_t(MemoryLayout<Int32>.size))
            }
        }

        var addr = sockaddr_in()
        addr.sin_family = sa_family_t(AF_INET)
        addr.sin_port = in_port_t(UInt16(0).bigEndian)
        addr.sin_addr = in_addr(s_addr: inet_addr("127.0.0.1"))

        let bindResult = withUnsafePointer(to: &addr) {
            $0.withMemoryRebound(to: sockaddr.self, capacity: 1) {
                testBind(listenFd, $0, socklen_t(MemoryLayout<sockaddr_in>.size))
            }
        }
        if bindResult != 0 {
            _ = testClose(listenFd)
            throw NSError(domain: "ScratchBirdTests", code: -1, userInfo: [NSLocalizedDescriptionKey: "bind() failed"])
        }

        if testListen(listenFd, 4) != 0 {
            _ = testClose(listenFd)
            throw NSError(domain: "ScratchBirdTests", code: -1, userInfo: [NSLocalizedDescriptionKey: "listen() failed"])
        }

        var boundAddr = sockaddr_in()
        var length = socklen_t(MemoryLayout<sockaddr_in>.size)
        let sockNameResult = withUnsafeMutablePointer(to: &boundAddr) {
            $0.withMemoryRebound(to: sockaddr.self, capacity: 1) {
                testGetSockName(listenFd, $0, &length)
            }
        }
        if sockNameResult != 0 {
            _ = testClose(listenFd)
            throw NSError(domain: "ScratchBirdTests", code: -1, userInfo: [NSLocalizedDescriptionKey: "getsockname() failed"])
        }
        self.port = Int(UInt16(bigEndian: boundAddr.sin_port))

        group.enter()
        DispatchQueue(label: "scratchbird.swift.auth.bootstrap.server").async {
            defer {
                _ = testClose(listenFd)
                self.group.leave()
            }

            let clientFd = testAccept(listenFd, nil, nil)
            if clientFd < 0 {
                self.scenarioError = NSError(
                    domain: "ScratchBirdTests",
                    code: -1,
                    userInfo: [NSLocalizedDescriptionKey: "accept() failed"]
                )
                return
            }

            defer {
                _ = testClose(clientFd)
            }

            do {
                try handler(clientFd)
            } catch {
                self.scenarioError = error
            }
        }
    }

    func wait() throws {
        group.wait()
        if let scenarioError {
            throw scenarioError
        }
    }
}

private func readProtocolMessage(from fd: Int32) throws -> ProtocolMessage {
    let headerData = try readExact(from: fd, length: headerSize)
    let header = try decodeHeader(headerData)
    let payload = header.length > 0 ? try readExact(from: fd, length: Int(header.length)) : Data()
    return ProtocolMessage(header: header, payload: payload)
}

private func writeProtocolMessage(
    to fd: Int32,
    type: MessageType,
    payload: Data,
    sequence: UInt32 = 0,
    attachmentId: Data = Data(repeating: 0, count: 16),
    txnId: UInt64 = 0
) throws {
    let data = encodeMessage(
        header: MessageHeader(
            type: type,
            flags: 0,
            length: UInt32(payload.count),
            sequence: sequence,
            attachmentId: attachmentId,
            txnId: txnId
        ),
        payload: payload
    )
    try writeAll(to: fd, data: data)
}

private func readManagerFrame(from fd: Int32) throws -> (type: UInt8, payload: Data) {
    let header = try readExact(from: fd, length: Int(testManagerHeaderSize))
    let magic = UInt32(littleEndian: header.subdata(in: 0..<4).withUnsafeBytes { $0.load(as: UInt32.self) })
    if magic != testManagerProtocolMagic {
        throw NSError(domain: "ScratchBirdTests", code: -1, userInfo: [NSLocalizedDescriptionKey: "Manager frame magic mismatch"])
    }
    let type = header[6]
    let payloadLength = UInt32(littleEndian: header.subdata(in: 8..<12).withUnsafeBytes { $0.load(as: UInt32.self) })
    let payload = payloadLength > 0 ? try readExact(from: fd, length: Int(payloadLength)) : Data()
    return (type, payload)
}

private func writeManagerFrame(to fd: Int32, type: UInt8, payload: Data) throws {
    var frame = Data()
    frame.append(contentsOf: withUnsafeBytes(of: testManagerProtocolMagic.littleEndian, Array.init))
    frame.append(contentsOf: withUnsafeBytes(of: testManagerProtocolVersion.littleEndian, Array.init))
    frame.append(type)
    frame.append(0)
    frame.append(contentsOf: withUnsafeBytes(of: UInt32(payload.count).littleEndian, Array.init))
    frame.append(payload)
    try writeAll(to: fd, data: frame)
}

private func readExact(from fd: Int32, length: Int) throws -> Data {
    var out = Data(count: length)
    var offset = 0
    while offset < length {
        let read = out.withUnsafeMutableBytes { ptr in
            testRecv(fd, ptr.baseAddress!.advanced(by: offset), length - offset, 0)
        }
        if read <= 0 {
            throw NSError(domain: "ScratchBirdTests", code: -1, userInfo: [NSLocalizedDescriptionKey: "recv() failed"])
        }
        offset += read
    }
    return out
}

private func writeAll(to fd: Int32, data: Data) throws {
    var offset = 0
    while offset < data.count {
        let written = data.withUnsafeBytes { ptr in
            testSend(fd, ptr.baseAddress!.advanced(by: offset), data.count - offset, 0)
        }
        if written <= 0 {
            throw NSError(domain: "ScratchBirdTests", code: -1, userInfo: [NSLocalizedDescriptionKey: "send() failed"])
        }
        offset += written
    }
}

private func makeAuthRequestPayload(method: UInt8) -> Data {
    Data([method, 0, 0, 0])
}

private func makeAuthContinuePayload(method: UInt8, stage: UInt8, value: String) -> Data {
    let bytes = Data(value.utf8)
    var payload = Data([method, stage, 0, 0])
    payload.append(contentsOf: withUnsafeBytes(of: UInt32(bytes.count).littleEndian, Array.init))
    payload.append(bytes)
    return payload
}

private func makeAuthOkPayload(serverInfo: String) -> Data {
    let serverInfoBytes = Data(serverInfo.utf8)
    var payload = Data(repeating: 0, count: 16)
    payload.append(contentsOf: withUnsafeBytes(of: UInt32(serverInfoBytes.count).littleEndian, Array.init))
    payload.append(serverInfoBytes)
    return payload
}

private func extractScramNonce(fromClientFirst clientFirst: String) -> String {
    for part in clientFirst.split(separator: ",") {
        if part.hasPrefix("r=") {
            return String(part.dropFirst(2))
        }
    }
    return ""
}

File: packaging/2026.07.03/drivers/dotnet/examples/tests/ScratchBird.Data.Tests/AuthBootstrapContractTests.cs
// Copyright (c) 2026 ScratchBird Software Inc.
//
// This Source Code Form is subject to the terms of the Mozilla Public
// License, v. 2.0. If a copy of the MPL was not distributed with this
// file, You can obtain one at https://mozilla.org/MPL/2.0/.
//
// SPDX-License-Identifier: MPL-2.0

using System.Buffers.Binary;
using System;
using System.IO;
using System.Net;
using System.Net.Sockets;
using System.Text;
using System.Threading.Tasks;
using ScratchBird.Data;
using Xunit;

namespace ScratchBird.Data.Tests;

public sealed class AuthBootstrapContractTests
{
    [Fact]
    public void ProbeAuthSurfaceDirectReportsScramSha512()
    {
        using var harness = LoopbackHarness.Start((stream, capture) =>
        {
            ReceiveProtocolMessage(stream);
            SendProtocolMessage(stream, BuildAuthRequestMessage(1, AuthMethod.SCRAM_SHA_512));
        });

        var result = ScratchBirdConnection.ProbeAuthSurface(
            $"Host=127.0.0.1;Port={harness.Port};Database=main;SSLMode=disable;AllowInsecure=true");

        Assert.True(result.Reachable);
        Assert.Equal("direct", result.FrontDoorMode);
        Assert.Equal((byte)AuthMethod.SCRAM_SHA_512, result.RequiredMethodCode);
        Assert.Equal("SCRAM_SHA_512", result.RequiredMethodName);
        Assert.Equal("scratchbird.auth.scram", result.RequiredAuthPluginId);
        Assert.Single(result.AdmittedMethods);
        Assert.True(result.AdmittedMethods[0].ExecutableLocally);
    }

    [Fact]
    public void ProbeAuthSurfaceManagerProxyReportsToken()
    {
        using var harness = LoopbackHarness.Start((stream, capture) =>
        {
            var hello = ReceiveManagerFrame(stream);
            Assert.Equal(0x65, hello.Type);
            SendManagerFrame(stream, 0x64, Array.Empty<byte>());

            var authStart = ReceiveManagerFrame(stream);
            Assert.Equal(0x66, authStart.Type);
            AssertManagerAuthStartPayload(authStart.Payload, expectInlineToken: false);

            SendManagerFrame(stream, 0x12, Array.Empty<byte>());
        });

        var result = ScratchBirdConnection.ProbeAuthSurface(
            $"Host=127.0.0.1;Port={harness.Port};Database=main;Username=admin;SSLMode=disable;AllowInsecure=true;Front_Door_Mode=manager_proxy");

        Assert.True(result.Reachable);
        Assert.Equal("manager_proxy", result.FrontDoorMode);
        Assert.Equal((byte)AuthMethod.TOKEN, result.RequiredMethodCode);
        Assert.Equal("TOKEN", result.RequiredMethodName);
        Assert.Equal("scratchbird.auth.token_authkey", result.RequiredAuthPluginId);
        Assert.Single(result.AdmittedMethods);
        Assert.True(result.AdmittedMethods[0].ExecutableLocally);
    }

    [Fact]
    public void HandshakeSupportsScramSha512AndResolvedContext()
    {
        using var harness = LoopbackHarness.Start((stream, capture) =>
        {
            ReceiveProtocolMessage(stream);
            SendProtocolMessage(stream, BuildAuthRequestMessage(1, AuthMethod.SCRAM_SHA_512));

            var clientFirst = Encoding.UTF8.GetString(ReceiveProtocolMessage(stream).Payload);
            var nonce = ExtractScramNonce(clientFirst);
            var serverFirst = $"r={nonce}server,s=QSXCR+Q6sek8bf92,i=4096";
            SendProtocolMessage(stream, BuildAuthContinueMessage(2, AuthMethod.SCRAM_SHA_512, 0, serverFirst));

            var clientFinal = Encoding.UTF8.GetString(ReceiveProtocolMessage(stream).Payload);
            Assert.Contains("c=biws", clientFinal, StringComparison.Ordinal);
            Assert.Contains("p=", clientFinal, StringComparison.Ordinal);

            SendProtocolMessage(stream, BuildAuthOkMessage(3));
            SendProtocolMessage(stream, BuildReadyMessage(4, 0));
        });

        using var connection = new ScratchBirdConnection(
            $"Host=127.0.0.1;Port={harness.Port};Database=main;Username=sb_admin;Password=SbAdmin_Compat1!;SSLMode=disable;AllowInsecure=true;Pooling=false");
        connection.Open();

        var resolved = connection.GetResolvedAuthContext();
        Assert.True(resolved.Attached);
        Assert.False(resolved.ManagerAuthenticated);
        Assert.Equal("direct", resolved.FrontDoorMode);
        Assert.Equal((byte)AuthMethod.SCRAM_SHA_512, resolved.ResolvedMethodCode);
        Assert.Equal("SCRAM_SHA_512", resolved.ResolvedMethodName);
        Assert.Equal("scratchbird.auth.scram", resolved.ResolvedAuthPluginId);
    }

    [Fact]
    public void HandshakeSupportsTokenAuth()
    {
        const string expectedToken = "oidc-bearer-token";

        using var harness = LoopbackHarness.Start((stream, captured) =>
        {
            ReceiveProtocolMessage(stream);
            SendProtocolMessage(stream, BuildAuthRequestMessage(1, AuthMethod.TOKEN));

            captured.TokenPayload = Encoding.UTF8.GetString(ReceiveProtocolMessage(stream).Payload);
            SendProtocolMessage(stream, BuildAuthOkMessage(2));
            SendProtocolMessage(stream, BuildReadyMessage(3, 0));
        });

        using var connection = new ScratchBirdConnection(
            $"Host=127.0.0.1;Port={harness.Port};Database=main;Username=sb_admin;Password=unused;SSLMode=disable;AllowInsecure=true;Pooling=false;Auth_Token={expectedToken}");
        connection.Open();

        Assert.Equal(expectedToken, harness.Captured.TokenPayload);
        var resolved = connection.GetResolvedAuthContext();
        Assert.Equal((byte)AuthMethod.TOKEN, resolved.ResolvedMethodCode);
        Assert.Equal("TOKEN", resolved.ResolvedMethodName);
        Assert.Equal("scratchbird.auth.token_authkey", resolved.ResolvedAuthPluginId);
    }

    [Fact]
    public void HandshakePeerFailsClosed()
    {
        using var harness = LoopbackHarness.Start((stream, capture) =>
        {
            ReceiveProtocolMessage(stream);
            SendProtocolMessage(stream, BuildAuthRequestMessage(1, AuthMethod.PEER));
        });

        var client = new ProtocolClient();
        var config = ScratchBirdConfig.FromConnectionString(
            $"Host=127.0.0.1;Port={harness.Port};Database=main;Username=sb_admin;Password=unused;SSLMode=disable;AllowInsecure=true;Pooling=false");

        var ex = Assert.Throws<ScratchBirdNotSupportedException>(() => client.Connect(config));
        Assert.Equal("0A000", ex.SqlState);
        Assert.Contains("PEER", ex.Message, StringComparison.Ordinal);
    }

    private static ProtocolMessage ReceiveProtocolMessage(NetworkStream stream)
    {
        var headerBytes = ReadExact(stream, ProtocolConstants.HeaderSize);
        var header = ProtocolMessage.ParseHeader(headerBytes);
        var payload = header.Length > 0 ? ReadExact(stream, (int)header.Length) : Array.Empty<byte>();
        return new ProtocolMessage(header, payload);
    }

    private static void SendProtocolMessage(NetworkStream stream, ProtocolMessage message)
    {
        var bytes = message.ToBytes();
        stream.Write(bytes, 0, bytes.Length);
        stream.Flush();
    }

    private static ProtocolMessage BuildAuthRequestMessage(uint sequence, AuthMethod method)
    {
        var payload = new byte[4];
        payload[0] = (byte)method;
        return new ProtocolMessage(
            new MessageHeader((byte)MessageType.AUTH_REQUEST, 0, (uint)payload.Length, sequence, new byte[16], 0),
            payload);
    }

    private static ProtocolMessage BuildAuthContinueMessage(uint sequence, AuthMethod method, byte stage, string data)
    {
        var dataBytes = Encoding.UTF8.GetBytes(data);
        var payload = new byte[8 + dataBytes.Length];
        payload[0] = (byte)method;
        payload[1] = stage;
        BinaryPrimitives.WriteUInt32LittleEndian(payload.AsSpan(4, 4), (uint)dataBytes.Length);
        Buffer.BlockCopy(dataBytes, 0, payload, 8, dataBytes.Length);
        return new ProtocolMessage(
            new MessageHeader((byte)MessageType.AUTH_CONTINUE, 0, (uint)payload.Length, sequence, new byte[16], 0),
            payload);
    }

    private static ProtocolMessage BuildAuthOkMessage(uint sequence)
    {
        var payload = new byte[20];
        BinaryPrimitives.WriteUInt32LittleEndian(payload.AsSpan(16, 4), 0);
        return new ProtocolMessage(
            new MessageHeader((byte)MessageType.AUTH_OK, 0, (uint)payload.Length, sequence, new byte[16], 0),
            payload);
    }

    private static ProtocolMessage BuildReadyMessage(uint sequence, ulong txnId)
    {
        var payload = new byte[20];
        payload[0] = 0;
        BinaryPrimitives.WriteUInt64LittleEndian(payload.AsSpan(4, 8), txnId);
        BinaryPrimitives.WriteUInt64LittleEndian(payload.AsSpan(12, 8), txnId);
        return new ProtocolMessage(
            new MessageHeader((byte)MessageType.READY, 0, (uint)payload.Length, sequence, new byte[16], txnId),
            payload);
    }

    private static byte[] ReadExact(NetworkStream stream, int length)
    {
        var buffer = new byte[length];
        var offset = 0;
        while (offset < length)
        {
            var read = stream.Read(buffer, offset, length - offset);
            if (read <= 0)
            {
                throw new IOException("Connection closed while reading test payload");
            }
            offset += read;
        }
        return buffer;
    }

    private static string ExtractScramNonce(string clientFirst)
    {
        const string marker = "r=";
        var index = clientFirst.IndexOf(marker, StringComparison.Ordinal);
        Assert.True(index >= 0, $"SCRAM client-first message missing nonce: {clientFirst}");
        return clientFirst[(index + marker.Length)..];
    }

    private static void AssertManagerAuthStartPayload(byte[] payload, bool expectInlineToken)
    {
        var offset = 0;
        var userLength = BinaryPrimitives.ReadUInt32LittleEndian(payload.AsSpan(offset, 4));
        offset += 4 + (int)userLength;
        Assert.True(offset < payload.Length);
        Assert.Equal(4, payload[offset]);
        offset += 1;
        var tokenLength = BinaryPrimitives.ReadUInt32LittleEndian(payload.AsSpan(offset, 4));
        Assert.Equal(expectInlineToken ? 1U : 0U, tokenLength > 0 ? 1U : 0U);
    }

    private static ManagerFrame ReceiveManagerFrame(NetworkStream stream)
    {
        var header = ReadExact(stream, 12);
        var magic = BinaryPrimitives.ReadUInt32LittleEndian(header.AsSpan(0, 4));
        Assert.Equal(0x42444253u, magic);
        var version = BinaryPrimitives.ReadUInt16LittleEndian(header.AsSpan(4, 2));
        Assert.Equal((ushort)0x0101, version);
        var type = header[6];
        var payloadLength = BinaryPrimitives.ReadUInt32LittleEndian(header.AsSpan(8, 4));
        var payload = payloadLength > 0 ? ReadExact(stream, (int)payloadLength) : Array.Empty<byte>();
        return new ManagerFrame(type, payload);
    }

    private static void SendManagerFrame(NetworkStream stream, byte type, byte[] payload)
    {
        var frame = new byte[12 + payload.Length];
        BinaryPrimitives.WriteUInt32LittleEndian(frame.AsSpan(0, 4), 0x42444253u);
        BinaryPrimitives.WriteUInt16LittleEndian(frame.AsSpan(4, 2), 0x0101);
        frame[6] = type;
        BinaryPrimitives.WriteUInt32LittleEndian(frame.AsSpan(8, 4), (uint)payload.Length);
        if (payload.Length > 0)
        {
            Buffer.BlockCopy(payload, 0, frame, 12, payload.Length);
        }
        stream.Write(frame, 0, frame.Length);
        stream.Flush();
    }

    private sealed record ManagerFrame(byte Type, byte[] Payload);

    private sealed class HarnessCapture
    {
        public string? TokenPayload { get; set; }
    }

    private sealed class LoopbackHarness : IDisposable
    {
        private readonly TcpListener _listener;
        private readonly Task _task;
        public int Port { get; }
        public HarnessCapture Captured { get; }

        private LoopbackHarness(TcpListener listener, Task task, int port, HarnessCapture captured)
        {
            _listener = listener;
            _task = task;
            Port = port;
            Captured = captured;
        }

        public static LoopbackHarness Start(Action<NetworkStream, HarnessCapture> handler)
        {
            var listener = new TcpListener(IPAddress.Loopback, 0);
            listener.Start();
            var port = ((IPEndPoint)listener.LocalEndpoint).Port;
            var capture = new HarnessCapture();
            var task = Task.Run(() =>
            {
                using var client = listener.AcceptTcpClient();
                using var stream = client.GetStream();
                handler(stream, capture);
            });

            return new LoopbackHarness(listener, task, port, capture);
        }

        public void Dispose()
        {
            _listener.Stop();
            _task.GetAwaiter().GetResult();
        }
    }
}
